contracts/examples/DXswapRelayer.sol

```diff
@@ -0,0 +1,392 @@
+
+import './OracleCreator.sol';
+import './../interfaces/IDXswapFactory.sol';
+import './../interfaces/IDXswapRouter.sol';
+import './../libraries/TransferHelper.sol';
+import './../interfaces/IERC20.sol';
+import './../interfaces/IWETH.sol';
+import './../libraries/SafeMath.sol';
+import './../libraries/DXswapLibrary.sol';
+
+contract DXswapRelayer {
+    using SafeMath for uint256;
+
+    event NewOrder(
+        uint256 indexed _orderIndex,
+        uint8 indexed _action
+    );
+
+    event ExecutedOrder(
+        uint256 indexed _orderIndex
+    );
+
+    event WithdrawnExpiredOrder(
+        uint256 indexed _orderIndex
+    );
+
+    struct Order {
+        uint8 action; // 1=provision; 2=removal
+        address tokenA;
+        address tokenB;
+        uint256 amountA;
+        uint256 amountB;
+        uint256 liquidity;
+        uint256 priceTolerance;
+        uint256 minReserveA;
+        uint256 minReserveB;
+        address oraclePair;
+        uint256 deadline;
+        uint256 maxWindowTime;
+        uint256 oracleId;
+        address factory;
+        bool executed;
+    }
+
+    uint256 public immutable GAS_ORACLE_UPDATE = 168364;
+    uint256 public immutable PARTS_PER_MILLION = 1000000;
+    uint256 public immutable BOUNTY = 0.01 ether; // To be decided
+    uint8 public immutable PROVISION = 1;
+    uint8 public immutable REMOVAL = 2;
+
+    address payable public immutable owner;
+    address public immutable dxSwapFactory;
+    address public immutable dxSwapRouter;
+    address public immutable uniswapFactory;
+    address public immutable uniswapRouter;
+    address public immutable WETH;
+
+    OracleCreator oracleCreator;
+    uint256 public orderCount;
+    mapping(uint256 => Order) orders;
+
+    constructor(
+        address payable _owner,
+        address _dxSwapFactory,
+        address _dxSwapRouter,
+        address _uniswapFactory,
+        address _uniswapRouter,
+        address _WETH,
+        OracleCreator _oracleCreater
+    ) public {
+        owner = _owner;
+        dxSwapFactory = _dxSwapFactory;
+        dxSwapRouter = _dxSwapRouter;
+        uniswapFactory = _uniswapFactory;
+        uniswapRouter = _uniswapRouter;
+        WETH = _WETH;
+        oracleCreator = _oracleCreater;
+    }
+
+    function orderLiquidityProvision(
+        address tokenA,
+        address tokenB,
+        uint256 amountA,
+        uint256 amountB,
+        uint256 priceTolerance,
+        uint256 minReserveA,
+        uint256 minReserveB,
+        uint256 maxWindowTime,
+        uint256 deadline,
+        address factory
+    ) external payable returns (uint256 orderIndex) {
+        require(factory == dxSwapFactory || factory == uniswapFactory, 'DXswapRelayer: INVALID_FACTORY');
+        require(msg.sender == owner, 'DXswapRelayer: CALLER_NOT_OWNER');
+        require(tokenA != tokenB, 'DXswapRelayer: INVALID_PAIR');
+        require(tokenA < tokenB, 'DXswapRelayer: INVALID_TOKEN_ORDER');
```

Comment:
What actually makes a token evaluate as less than another?  Will people be able to figure this out relatively easily?

Reply:
@jpkcambridge it's actually just a string sorting. This is indeed a little UX nightmare in our current integration as we plan it. I would expect the frontend that speaks to the contract to actually sort the tokens under the hood, so the user interacting with the contract doesn't have to care about it – unfortunetly Multicall Scheme doesn't have such capabilities as of now. In the first place we need to provide a good documentation and should find a solution to solve it on the Multicall UI side of things